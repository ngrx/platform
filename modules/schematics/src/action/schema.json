{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxAction",
  "title": "NgRx Action Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the action.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What should be the name of the action?"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "skipTests": {
      "type": "boolean",
      "description": "When true, does not create test files.",
      "default": false
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description": "Group actions file within 'actions' folder",
      "aliases": ["g"]
    },
    "api": {
      "type": "boolean",
      "default": false,
      "description": "Specifies if api success and failure actions should be generated.",
      "aliases": ["a"],
      "x-prompt": "Should we generate success and failure actions?"
    },
    "creators": {
      "type": "boolean",
      "default": true,
      "description": "Specifies whether to use creator functions for handling actions and reducers.",
      "aliases": ["c"],
      "x-prompt": "Do you want to use the create function?"
    }
  },
  "required": []
}
