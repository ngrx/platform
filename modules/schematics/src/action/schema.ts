export interface Schema {
  /**
   * The name of the component.
   */

  name: string;
  /**
   * The path to create the component.
   */

  path?: string;
  /**
   * The name of the project.
   */
  project?: string;
  /**
   * Specifies if a spec file is generated.
   */
  spec?: boolean;
  /**
   * Flag to indicate if a dir is created.
   */

  flat?: boolean;
  /**
   * Group actions file within 'actions' folder
   */

  group?: boolean;
}
