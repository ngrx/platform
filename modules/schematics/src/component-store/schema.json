{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxComponentStore",
  "title": "NgRx Component Store Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the component store.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What should be the name of the component store?"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component store.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "skipTests": {
      "type": "boolean",
      "default": false,
      "description": "When true, does not create test files."
    },
    "component": {
      "type": "string",
      "default": "",
      "description": "Allows specification of the declaring component.",
      "alias": "m",
      "subtype": "filepath",
      "x-prompt": "To which component (path) should the component store be provided in?"
    },
    "module": {
      "type": "string",
      "default": "",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath",
      "x-prompt": "To which module (path) should the component store be provided in?"
    }
  },
  "required": []
}
