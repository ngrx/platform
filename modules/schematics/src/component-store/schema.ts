export interface Schema {
  /**
   * The name of the component store.
   */
  name: string;

  /**
   * The path to create the component store.
   */
  path?: string;

  /**
   * The name of the project.
   */
  project?: string;

  /**
   * Flag to indicate if a dir is created.
   */
  flat?: boolean;

  /**
   * When true, does not create test files.
   */
  skipTests?: boolean;

  /**
   * Allows specification of the declaring component.
   */
  component?: string;

  /**
   * Allows specification of the declaring module.
   */
  module?: string;
}
