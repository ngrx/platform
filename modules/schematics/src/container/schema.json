{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxContainer",
  "title": "NgRx Container Options Schema",
  "type": "object",
  "properties": {
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "name": {
      "type": "string",
      "description": "The name of the component.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What should be the name of the container component?"
    },
    "inlineStyle": {
      "description": "Specifies if the style will be in the ts file.",
      "type": "boolean",
      "alias": "s"
    },
    "inlineTemplate": {
      "description": "Specifies if the template will be in the ts file.",
      "type": "boolean",
      "alias": "t"
    },
    "viewEncapsulation": {
      "description": "Specifies the view encapsulation strategy.",
      "enum": ["Emulated", "Native", "None"],
      "type": "string",
      "alias": "v"
    },
    "changeDetection": {
      "description": "Specifies the change detection strategy.",
      "enum": ["Default", "OnPush"],
      "type": "string",
      "alias": "c"
    },
    "prefix": {
      "type": "string",
      "format": "html-selector",
      "description": "The prefix to apply to generated selectors.",
      "alias": "p"
    },
    "style": {
      "description": "The file extension or preprocessor to use for style files.",
      "type": "string"
    },
    "skipTests": {
      "type": "boolean",
      "description": "When true, does not create test files."
    },
    "flat": {
      "type": "boolean",
      "description": "Flag to indicate if a dir is created."
    },
    "skipImport": {
      "type": "boolean",
      "description": "Flag to skip the module import."
    },
    "selector": {
      "type": "string",
      "format": "html-selector",
      "description": "The selector to use for the component."
    },
    "module": {
      "type": "string",
      "description": "Allows specification of the declaring module.",
      "alias": "m"
    },
    "export": {
      "type": "boolean",
      "description": "Specifies if declaring module exports the component."
    },
    "state": {
      "type": "string",
      "description": "Specifies the path to the state exports."
    },
    "stateInterface": {
      "type": "string",
      "default": "State",
      "description": "Specifies the interface for the state."
    },
    "testDepth": {
      "description": "Specifies whether to create a unit test or an integration test.",
      "enum": ["unit", "integration"],
      "type": "string",
      "default": "integration"
    }
  },
  "required": []
}
