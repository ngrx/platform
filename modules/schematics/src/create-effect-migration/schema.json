{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxCreateEffectMigration",
  "title": "NgRx Effect Migration from @Effect to createEffect",
  "type": "object",
  "properties": {},
  "required": []
}
