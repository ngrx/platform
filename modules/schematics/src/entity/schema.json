{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxEntity",
  "title": "NgRx Entity Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the entity.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What should be the name of the entity?"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "skipTests": {
      "type": "boolean",
      "description": "When true, does not create test files.",
      "default": false
    },
    "reducers": {
      "type": "string",
      "description": "Specifies the reducers file.",
      "aliases": ["r"]
    },
    "module": {
      "type": "string",
      "description": "Specifies the declaring module.",
      "aliases": ["m"]
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description": "Group actions, reducers and effects within relative subfolders",
      "aliases": ["g"]
    },
    "creators": {
      "type": "boolean",
      "default": true,
      "description": "Specifies whether to use creator functions for handling actions and reducers.",
      "aliases": ["c"],
      "x-prompt": "Do you want to use the create function?"
    }
  },
  "required": []
}
