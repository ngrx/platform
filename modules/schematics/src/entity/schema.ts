export interface Schema {
  /**
   * The name of the component.
   */

  name: string;
  /**
   * The path to create the effect.
   */

  path?: string;
  /**
   * The name of the project.
   */
  project?: string;
  /**
   * Flag to indicate if a dir is created.
   */
  flat?: boolean;
  /**
   * Specifies if a spec file is generated.
   */
  spec?: boolean;
  /**
   * Allows specification of the declaring module.
   */

  module?: string;
  /**
   * Allows specification of the declaring reducers.
   */

  reducers?: string;
  /**
   * Specifies if this is grouped within sub folders
   */

  group?: boolean;
}
