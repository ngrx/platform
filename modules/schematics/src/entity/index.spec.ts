import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { Schema as EntityOptions } from './schema';
import {
  getTestProjectPath,
  createWorkspace,
  defaultWorkspaceOptions,
  defaultAppOptions,
} from '@ngrx/schematics-core/testing';

describe('Entity Schematic', () => {
  const schematicRunner = new SchematicTestRunner(
    '@ngrx/schematics',
    path.join(__dirname, '../../collection.json')
  );
  const defaultOptions: EntityOptions = {
    name: 'foo',
    project: 'bar',
  };

  const projectPath = getTestProjectPath();

  let appTree: UnitTestTree;

  beforeEach(async () => {
    appTree = await createWorkspace(schematicRunner, appTree);
  });

  it('should create 3 files', async () => {
    const tree = await schematicRunner.runSchematic(
      'entity',
      defaultOptions,
      appTree
    );

    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo.actions.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo.model.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo.reducer.ts`)
    ).toBeGreaterThanOrEqual(0);

    expect(
      tree.readContent(`${projectPath}/src/app/foo.actions.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/foo.model.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/foo.reducer.ts`)
    ).toMatchSnapshot();
  });

  it('should create 3 files of an entity to specified project if provided', async () => {
    const options = {
      ...defaultOptions,
      project: 'baz',
    };

    const specifiedProjectPath = getTestProjectPath(defaultWorkspaceOptions, {
      ...defaultAppOptions,
      name: 'baz',
    });

    const tree = await schematicRunner.runSchematic('entity', options, appTree);
    const files = tree.files;
    expect(
      files.indexOf(`${specifiedProjectPath}/src/lib/foo.actions.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${specifiedProjectPath}/src/lib/foo.model.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${specifiedProjectPath}/src/lib/foo.reducer.ts`)
    ).toBeGreaterThanOrEqual(0);

    expect(
      tree.readContent(`${projectPath}/src/lib/foo.actions.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/lib/foo.model.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/lib/foo.reducer.ts`)
    ).toMatchSnapshot();
  });

  it('should create a folder if flat is false', async () => {
    const tree = await schematicRunner.runSchematic(
      'entity',
      {
        ...defaultOptions,
        flat: false,
      },
      appTree
    );
    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo/foo.actions.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo/foo.model.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo/foo.reducer.ts`)
    ).toBeGreaterThanOrEqual(0);
  });

  it('should create 4 files if spec is true', async () => {
    const options = {
      ...defaultOptions,
    };
    const tree = await schematicRunner.runSchematic('entity', options, appTree);

    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo.actions.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo.model.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo.reducer.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      tree.files.indexOf(`${projectPath}/src/app/foo.reducer.spec.ts`)
    ).toBeGreaterThanOrEqual(0);
  });

  it('should import into a specified module', async () => {
    const options = { ...defaultOptions, module: 'app.module.ts' };

    const tree = await schematicRunner.runSchematic('entity', options, appTree);
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);

    expect(content).toMatchSnapshot();
  });
  it('should create all files of an entity within grouped and nested folders', async () => {
    const options = { ...defaultOptions, flat: false, group: true };

    const tree = await schematicRunner.runSchematic('entity', options, appTree);
    const files = tree.files;

    expect(
      files.indexOf(`${projectPath}/src/app/foo/actions/foo.actions.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/foo/models/foo.model.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/foo/reducers/foo.reducer.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/foo/reducers/foo.reducer.spec.ts`)
    ).toBeGreaterThanOrEqual(0);

    expect(
      tree.readContent(`${projectPath}/src/app/foo/actions/foo.actions.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/foo/models/foo.model.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/foo/reducers/foo.reducer.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(
        `${projectPath}/src/app/foo/reducers/foo.reducer.spec.ts`
      )
    ).toMatchSnapshot();
  });

  it('should create all files of an entity within grouped folders if group is set', async () => {
    const options = { ...defaultOptions, group: true };

    const tree = await schematicRunner.runSchematic('entity', options, appTree);
    const files = tree.files;

    expect(
      files.indexOf(`${projectPath}/src/app/actions/foo.actions.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/models/foo.model.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/reducers/foo.reducer.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/reducers/foo.reducer.spec.ts`)
    ).toBeGreaterThanOrEqual(0);

    expect(
      tree.readContent(`${projectPath}/src/app/actions/foo.actions.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/models/foo.model.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/reducers/foo.reducer.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/reducers/foo.reducer.spec.ts`)
    ).toMatchSnapshot();
  });

  it('should update the state to plural', async () => {
    const options = {
      ...defaultOptions,
      name: 'user',
      reducers: 'reducers/index.ts',
    };

    const _reducerTree = await schematicRunner.runSchematic(
      'store',
      options,
      appTree
    );
    const tree = await schematicRunner.runSchematic('entity', options, appTree);
    const files = tree.files;
    expect(
      files.indexOf(`${projectPath}/src/app/user.actions.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/user.model.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/user.reducer.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/user.reducer.spec.ts`)
    ).toBeGreaterThanOrEqual(0);

    expect(
      tree.readContent(`${projectPath}/src/app/user.actions.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/user.model.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/user.reducer.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/user.reducer.spec.ts`)
    ).toMatchSnapshot();
    expect(
      tree.readContent(`${projectPath}/src/app/reducers/index.ts`)
    ).toMatchSnapshot();
  });
});
