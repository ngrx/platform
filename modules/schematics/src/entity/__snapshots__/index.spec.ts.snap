// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity Schematic should create 3 files 1`] = `
"import { createActionGroup, emptyProps, props } from '@ngrx/store';
import { Update } from '@ngrx/entity';

import { Foo } from './foo.model';

export const FooActions = createActionGroup({
  source: 'Foo/API',
  events: {
    'Load Foos': props<{ foos: Foo[] }>(),
    'Add Foo': props<{ foo: Foo }>(),
    'Upsert Foo': props<{ foo: Foo }>(),
    'Add Foos': props<{ foo: Foo[] }>(),
    'Upsert Foos': props<{ foo: Foo[] }>(),
    'Update Foo': props<{ foo: Update<Foo> }>(),
    'Update Foos': props<{ foos: Update<Foo>[] }>(),
    'Delete Foo': props<{ id: string }>(),
    'Delete Foos': props<{ ids: string[] }>(),
    'Clear Foos': emptyProps(),
  }
});
"
`;

exports[`Entity Schematic should create 3 files 2`] = `
"export interface Foo {
  id: string;
}
"
`;

exports[`Entity Schematic should create 3 files 3`] = `
"import { createFeature, createReducer, on } from '@ngrx/store';
import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { Foo } from './foo.model';
import { FooActions } from './foo.actions';

export const foosFeatureKey = 'foos';

export interface State extends EntityState<Foo> {
  // additional entities state properties
}

export const adapter: EntityAdapter<Foo> = createEntityAdapter<Foo>();

export const initialState: State = adapter.getInitialState({
  // additional entity state properties
});

export const reducer = createReducer(
  initialState,
  on(FooActions.addFoo,
    (state, action) => adapter.addOne(action.foo, state)
  ),
  on(FooActions.upsertFoo,
    (state, action) => adapter.upsertOne(action.foo, state)
  ),
  on(FooActions.addFoos,
    (state, action) => adapter.addMany(action.foos, state)
  ),
  on(FooActions.upsertFoos,
    (state, action) => adapter.upsertMany(action.foos, state)
  ),
  on(FooActions.updateFoo,
    (state, action) => adapter.updateOne(action.foo, state)
  ),
  on(FooActions.updateFoos,
    (state, action) => adapter.updateMany(action.foos, state)
  ),
  on(FooActions.deleteFoo,
    (state, action) => adapter.removeOne(action.id, state)
  ),
  on(FooActions.deleteFoos,
    (state, action) => adapter.removeMany(action.ids, state)
  ),
  on(FooActions.loadFoos,
    (state, action) => adapter.setAll(action.foos, state)
  ),
  on(FooActions.clearFoos,
    state => adapter.removeAll(state)
  ),
);

export const foosFeature = createFeature({
  name: foosFeatureKey,
  reducer,
  extraSelectors: ({ selectFoosState }) => ({
    ...adapter.getSelectors(selectFoosState)
  }),
});

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = foosFeature;
"
`;

exports[`Entity Schematic should create 3 files of an entity to specified project if provided 1`] = `""`;

exports[`Entity Schematic should create 3 files of an entity to specified project if provided 2`] = `""`;

exports[`Entity Schematic should create 3 files of an entity to specified project if provided 3`] = `""`;

exports[`Entity Schematic should create all files of an entity within grouped and nested folders 1`] = `
"import { createActionGroup, emptyProps, props } from '@ngrx/store';
import { Update } from '@ngrx/entity';

import { Foo } from '../../models/foo/foo.model';

export const FooActions = createActionGroup({
  source: 'Foo/API',
  events: {
    'Load Foos': props<{ foos: Foo[] }>(),
    'Add Foo': props<{ foo: Foo }>(),
    'Upsert Foo': props<{ foo: Foo }>(),
    'Add Foos': props<{ foo: Foo[] }>(),
    'Upsert Foos': props<{ foo: Foo[] }>(),
    'Update Foo': props<{ foo: Update<Foo> }>(),
    'Update Foos': props<{ foos: Update<Foo>[] }>(),
    'Delete Foo': props<{ id: string }>(),
    'Delete Foos': props<{ ids: string[] }>(),
    'Clear Foos': emptyProps(),
  }
});
"
`;

exports[`Entity Schematic should create all files of an entity within grouped and nested folders 2`] = `
"export interface Foo {
  id: string;
}
"
`;

exports[`Entity Schematic should create all files of an entity within grouped and nested folders 3`] = `
"import { createFeature, createReducer, on } from '@ngrx/store';
import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { Foo } from '../../models/foo/foo.model';
import { FooActions } from '../../actions/foo/foo.actions';

export const foosFeatureKey = 'foos';

export interface State extends EntityState<Foo> {
  // additional entities state properties
}

export const adapter: EntityAdapter<Foo> = createEntityAdapter<Foo>();

export const initialState: State = adapter.getInitialState({
  // additional entity state properties
});

export const reducer = createReducer(
  initialState,
  on(FooActions.addFoo,
    (state, action) => adapter.addOne(action.foo, state)
  ),
  on(FooActions.upsertFoo,
    (state, action) => adapter.upsertOne(action.foo, state)
  ),
  on(FooActions.addFoos,
    (state, action) => adapter.addMany(action.foos, state)
  ),
  on(FooActions.upsertFoos,
    (state, action) => adapter.upsertMany(action.foos, state)
  ),
  on(FooActions.updateFoo,
    (state, action) => adapter.updateOne(action.foo, state)
  ),
  on(FooActions.updateFoos,
    (state, action) => adapter.updateMany(action.foos, state)
  ),
  on(FooActions.deleteFoo,
    (state, action) => adapter.removeOne(action.id, state)
  ),
  on(FooActions.deleteFoos,
    (state, action) => adapter.removeMany(action.ids, state)
  ),
  on(FooActions.loadFoos,
    (state, action) => adapter.setAll(action.foos, state)
  ),
  on(FooActions.clearFoos,
    state => adapter.removeAll(state)
  ),
);

export const foosFeature = createFeature({
  name: foosFeatureKey,
  reducer,
  extraSelectors: ({ selectFoosState }) => ({
    ...adapter.getSelectors(selectFoosState)
  }),
});

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = foosFeature;
"
`;

exports[`Entity Schematic should create all files of an entity within grouped and nested folders 4`] = `
"import { reducer, initialState } from '../../reducers/foo/foo.reducer';

describe('Foo Reducer', () => {
  describe('unknown action', () => {
    it('should return the previous state', () => {
      const action = {} as any;

      const result = reducer(initialState, action);

      expect(result).toBe(initialState);
    });
  });
});
"
`;

exports[`Entity Schematic should create all files of an entity within grouped folders if group is set 1`] = `
"import { createActionGroup, emptyProps, props } from '@ngrx/store';
import { Update } from '@ngrx/entity';

import { Foo } from '../models/foo.model';

export const FooActions = createActionGroup({
  source: 'Foo/API',
  events: {
    'Load Foos': props<{ foos: Foo[] }>(),
    'Add Foo': props<{ foo: Foo }>(),
    'Upsert Foo': props<{ foo: Foo }>(),
    'Add Foos': props<{ foo: Foo[] }>(),
    'Upsert Foos': props<{ foo: Foo[] }>(),
    'Update Foo': props<{ foo: Update<Foo> }>(),
    'Update Foos': props<{ foos: Update<Foo>[] }>(),
    'Delete Foo': props<{ id: string }>(),
    'Delete Foos': props<{ ids: string[] }>(),
    'Clear Foos': emptyProps(),
  }
});
"
`;

exports[`Entity Schematic should create all files of an entity within grouped folders if group is set 2`] = `
"export interface Foo {
  id: string;
}
"
`;

exports[`Entity Schematic should create all files of an entity within grouped folders if group is set 3`] = `
"import { createFeature, createReducer, on } from '@ngrx/store';
import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { Foo } from '../models/foo.model';
import { FooActions } from '../actions/foo.actions';

export const foosFeatureKey = 'foos';

export interface State extends EntityState<Foo> {
  // additional entities state properties
}

export const adapter: EntityAdapter<Foo> = createEntityAdapter<Foo>();

export const initialState: State = adapter.getInitialState({
  // additional entity state properties
});

export const reducer = createReducer(
  initialState,
  on(FooActions.addFoo,
    (state, action) => adapter.addOne(action.foo, state)
  ),
  on(FooActions.upsertFoo,
    (state, action) => adapter.upsertOne(action.foo, state)
  ),
  on(FooActions.addFoos,
    (state, action) => adapter.addMany(action.foos, state)
  ),
  on(FooActions.upsertFoos,
    (state, action) => adapter.upsertMany(action.foos, state)
  ),
  on(FooActions.updateFoo,
    (state, action) => adapter.updateOne(action.foo, state)
  ),
  on(FooActions.updateFoos,
    (state, action) => adapter.updateMany(action.foos, state)
  ),
  on(FooActions.deleteFoo,
    (state, action) => adapter.removeOne(action.id, state)
  ),
  on(FooActions.deleteFoos,
    (state, action) => adapter.removeMany(action.ids, state)
  ),
  on(FooActions.loadFoos,
    (state, action) => adapter.setAll(action.foos, state)
  ),
  on(FooActions.clearFoos,
    state => adapter.removeAll(state)
  ),
);

export const foosFeature = createFeature({
  name: foosFeatureKey,
  reducer,
  extraSelectors: ({ selectFoosState }) => ({
    ...adapter.getSelectors(selectFoosState)
  }),
});

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = foosFeature;
"
`;

exports[`Entity Schematic should create all files of an entity within grouped folders if group is set 4`] = `
"import { reducer, initialState } from '../reducers/foo.reducer';

describe('Foo Reducer', () => {
  describe('unknown action', () => {
    it('should return the previous state', () => {
      const action = {} as any;

      const result = reducer(initialState, action);

      expect(result).toBe(initialState);
    });
  });
});
"
`;

exports[`Entity Schematic should import into a specified module 1`] = `
"import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

import { AppComponent } from './app.component';
import { StoreModule } from '@ngrx/store';
import * as fromFoo from './foo.reducer';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    StoreModule.forFeature(fromFoo.foosFeatureKey, fromFoo.reducer)
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
"
`;

exports[`Entity Schematic should update the state to plural 1`] = `
"import { createActionGroup, emptyProps, props } from '@ngrx/store';
import { Update } from '@ngrx/entity';

import { User } from './user.model';

export const UserActions = createActionGroup({
  source: 'User/API',
  events: {
    'Load Users': props<{ users: User[] }>(),
    'Add User': props<{ user: User }>(),
    'Upsert User': props<{ user: User }>(),
    'Add Users': props<{ user: User[] }>(),
    'Upsert Users': props<{ user: User[] }>(),
    'Update User': props<{ user: Update<User> }>(),
    'Update Users': props<{ users: Update<User>[] }>(),
    'Delete User': props<{ id: string }>(),
    'Delete Users': props<{ ids: string[] }>(),
    'Clear Users': emptyProps(),
  }
});
"
`;

exports[`Entity Schematic should update the state to plural 2`] = `
"export interface User {
  id: string;
}
"
`;

exports[`Entity Schematic should update the state to plural 3`] = `
"import { createFeature, createReducer, on } from '@ngrx/store';
import { EntityState, EntityAdapter, createEntityAdapter } from '@ngrx/entity';
import { User } from './user.model';
import { UserActions } from './user.actions';

export const usersFeatureKey = 'users';

export interface State extends EntityState<User> {
  // additional entities state properties
}

export const adapter: EntityAdapter<User> = createEntityAdapter<User>();

export const initialState: State = adapter.getInitialState({
  // additional entity state properties
});

export const reducer = createReducer(
  initialState,
  on(UserActions.addUser,
    (state, action) => adapter.addOne(action.user, state)
  ),
  on(UserActions.upsertUser,
    (state, action) => adapter.upsertOne(action.user, state)
  ),
  on(UserActions.addUsers,
    (state, action) => adapter.addMany(action.users, state)
  ),
  on(UserActions.upsertUsers,
    (state, action) => adapter.upsertMany(action.users, state)
  ),
  on(UserActions.updateUser,
    (state, action) => adapter.updateOne(action.user, state)
  ),
  on(UserActions.updateUsers,
    (state, action) => adapter.updateMany(action.users, state)
  ),
  on(UserActions.deleteUser,
    (state, action) => adapter.removeOne(action.id, state)
  ),
  on(UserActions.deleteUsers,
    (state, action) => adapter.removeMany(action.ids, state)
  ),
  on(UserActions.loadUsers,
    (state, action) => adapter.setAll(action.users, state)
  ),
  on(UserActions.clearUsers,
    state => adapter.removeAll(state)
  ),
);

export const usersFeature = createFeature({
  name: usersFeatureKey,
  reducer,
  extraSelectors: ({ selectUsersState }) => ({
    ...adapter.getSelectors(selectUsersState)
  }),
});

export const {
  selectIds,
  selectEntities,
  selectAll,
  selectTotal,
} = usersFeature;
"
`;

exports[`Entity Schematic should update the state to plural 4`] = `
"import { reducer, initialState } from './user.reducer';

describe('User Reducer', () => {
  describe('unknown action', () => {
    it('should return the previous state', () => {
      const action = {} as any;

      const result = reducer(initialState, action);

      expect(result).toBe(initialState);
    });
  });
});
"
`;

exports[`Entity Schematic should update the state to plural 5`] = `
"import { isDevMode } from '@angular/core';
import {
  ActionReducer,
  ActionReducerMap,
  createFeatureSelector,
  createSelector,
  MetaReducer
} from '@ngrx/store';
import * as fromUser from '../user.reducer';

export const userFeatureKey = 'user';

export interface State {

  [fromUser.usersFeatureKey]: fromUser.State;
}

export const reducers: ActionReducerMap<State> = {

  [fromUser.usersFeatureKey]: fromUser.reducer,
};


export const metaReducers: MetaReducer<State>[] = isDevMode() ? [] : [];
"
`;
